File: src/services/payment/upiIntent.ts
import { Linking, Platform } from 'react-native';
import QRCode from 'qrcode';
import { UPIPaymentConfig } from '@/types';

/**
 * Builds standard OS-level UPI intent deep link URI
 * Scheme: upi://pay?pa=...&pn=...&am=...&cu=INR&tn=...
 * Note is sanitized to alphanumeric characters only to prevent bank risk policy blocks.
 */
export function buildUPIIntentURL(config: UPIPaymentConfig): string {
  const { vpaId = '', payeeName = '', amount = 0, currency = 'INR', note = '' } = config;

  const formattedAmount = Number(amount || 0).toFixed(2);
  const cleanName = String(payeeName || 'Payee').trim();
  const encodedName = encodeURIComponent(cleanName);
  const cleanNote = String(note || 'FairShare')
    .replace(/[^a-zA-Z0-9 ]/g, ' ')
    .replace(/\s+/g, ' ')
    .trim() || 'FairShare';
  const encodedNote = encodeURIComponent(cleanNote);
  const cleanVpa = String(vpaId || '').trim();

  return `upi://pay?pa=${cleanVpa}&pn=${encodedName}&am=${formattedAmount}&cu=${currency}&tn=${encodedNote}`;
}

/**
 * Generates a pure boolean matrix for rendering UPI QR codes natively in React Native
 * without any HTML canvas or DOM dependencies.
 */
export function generateUPIMatrix(config: UPIPaymentConfig): { matrix: boolean[][]; moduleSize: number } {
  const url = buildUPIIntentURL(config);
  const qr = QRCode.create(url, { errorCorrectionLevel: 'M' });
  const moduleSize = qr.modules.size;
  const data = qr.modules.data;
  const matrix: boolean[][] = [];

  for (let r = 0; r < moduleSize; r++) {
    const row: boolean[] = [];
    for (let c = 0; c < moduleSize; c++) {
      row.push(Boolean(data[r * moduleSize + c]));
    }
    matrix.push(row);
  }
  return { matrix, moduleSize };
}

/**
 * Generates a base64 PNG data URL for a scannable UPI QR code (Node / Web environments).
 */
export async function generateUPIQRCode(config: UPIPaymentConfig): Promise<string> {
  const url = buildUPIIntentURL(config);
  return QRCode.toDataURL(url, {
    width: 256,
    margin: 1,
    color: {
      dark: '#000000',
      light: '#FFFFFF',
    },
  });
}

/**
 * Programmatically triggers OS-level UPI intent redirection
 * Redirects user to GPay, PhonePe, Paytm, or preferred installed UPI app.
 */
export async function launchUPIIntent(config: UPIPaymentConfig): Promise<{ success: boolean; message?: string }> {
  if (!config.vpaId || !config.vpaId.includes('@')) {
    return {
      success: false,
      message: 'Invalid Payee UPI ID (VPA). Please update profile with a valid UPI ID (e.g. name@upi).',
    };
  }

  const url = buildUPIIntentURL(config);

  try {
    const supported = await Linking.canOpenURL(url);

    if (supported || Platform.OS === 'android') {
      await Linking.openURL(url);
      return { success: true };
    } else {
      return {
        success: false,
        message: 'No supported UPI app found. Please copy the VPA ID to pay manually.',
      };
    }
  } catch (err: any) {
    console.warn('Error launching UPI Intent:', err);
    return {
      success: false,
      message: err?.message || 'Could not launch UPI payment app.',
    };
  }
}

File: lib/core/utils/upi_intent.dart
import 'package:url_launcher/url_launcher.dart';

/// Configuration for generating UPI payment intents
class UPIPaymentConfig {
  final String vpaId; // e.g. name@okaxis
  final String payeeName;
  final double amount;
  final String currency;
  final String note;
  final String? transactionRef;

  const UPIPaymentConfig({
    required this.vpaId,
    required this.payeeName,
    required this.amount,
    this.currency = 'INR',
    this.note = 'FairShare',
    this.transactionRef,
  });
}

/// Popular UPI payment apps on Android
enum UPIApp {
  generic('Any UPI App', null),
  gpay('Google Pay', 'com.google.android.apps.nbu.paisa.user'),
  phonepe('PhonePe', 'com.phonepe.app'),
  paytm('Paytm', 'net.one97.paytm'),
  cred('CRED', 'com.dreamplug.androidapp'),
  bhim('BHIM', 'in.org.npci.upiapp');

  final String displayName;
  final String? packageName;

  const UPIApp(this.displayName, this.packageName);
}

class UPIIntentResult {
  final bool success;
  final String? message;

  const UPIIntentResult({required this.success, this.message});
}

/// UPIIntentHelper builds and triggers standard OS-level UPI intent deep links.
class UPIIntentHelper {
  /// Builds standard OS-level UPI intent deep link URI
  /// Format: upi://pay?pa=...&pn=...&am=...&cu=INR&tn=...
  /// The note is sanitized to alphanumeric characters and spaces only to prevent bank risk policy blocks.
  static String buildUPIIntentURL(UPIPaymentConfig config) {
    final cleanVpa = config.vpaId.trim();
    final cleanName = config.payeeName.trim().isEmpty ? 'Payee' : config.payeeName.trim();
    final encodedName = Uri.encodeComponent(cleanName);
    final formattedAmount = config.amount.toStringAsFixed(2);

    String rawNote = config.note.trim().isEmpty ? 'FairShare' : config.note;
    String cleanNote = rawNote
        .replaceAll(RegExp(r'[^a-zA-Z0-9 ]'), ' ')
        .replaceAll(RegExp(r'\s+'), ' ')
        .trim();
    if (cleanNote.isEmpty) cleanNote = 'FairShare';
    final encodedNote = Uri.encodeComponent(cleanNote);

    return 'upi://pay?pa=$cleanVpa&pn=$encodedName&am=$formattedAmount&cu=${config.currency}&tn=$encodedNote';
  }

  /// Programmatically triggers OS-level UPI intent redirection.
  /// Launches user's installed UPI app (Google Pay, PhonePe, Paytm, CRED, etc.).
  static Future<UPIIntentResult> launchUPIIntent(UPIPaymentConfig config) async {
    final vpa = config.vpaId.trim();
    if (!vpa.contains('@') || vpa.length < 3) {
      return const UPIIntentResult(
        success: false,
        message: 'Invalid Payee UPI ID (VPA). Please update profile with a valid UPI ID (e.g. name@upi).',
      );
    }

    final urlString = buildUPIIntentURL(config);
    final uri = Uri.parse(urlString);

    try {
      final launched = await launchUrl(
        uri,
        mode: LaunchMode.externalNonBrowserApplication,
      );

      if (launched) {
        return const UPIIntentResult(success: true);
      } else {
        return const UPIIntentResult(
          success: false,
          message: 'No supported UPI app found. Please copy the UPI ID to pay manually.',
        );
      }
    } catch (e) {
      return UPIIntentResult(
        success: false,
        message: 'Could not launch UPI payment app: $e',
      );
    }
  }
}
